dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java

```diff
@@ -142,24 +161,157 @@ public void put(String path, FileStatus meta) {
    * @return the removed file meta, if exists
    */
   public Optional<FileStatus> removeFromMetaStore(String path) {
-    Optional<FileStatus> status = mMetastore.getDoraMeta(path);
+    invalidateListingCache(getPathParent(path));
+    Optional<FileStatus> status = mMetaStore.getDoraMeta(path);
     if (status.isPresent()) {
-      mMetastore.removeDoraMeta(path);
-      invalidateCachedFile(path, status.get().getFileInfo().getLength());
+      mMetaStore.removeDoraMeta(path);
     }
+    invalidateCachedFile(path);
     return status;
   }
 
-  private void invalidateCachedFile(String path, long length) {
-    FileId fileId = FileId.of(AlluxioURI.hash(path));
-    mCacheManager.deleteFile(fileId.toString());
-    for (PageId page: mCacheManager.getCachedPageIdsByFileId(fileId.toString(), length)) {
-      mCacheManager.delete(page);
+  /**
+   * Invalidates the listing cache of a given path.
+   * @param path the full ufs path
+   */
+  public void invalidateListingCache(String path) {
+    if (mListStatusCache != null) {
+      mListStatusCache.invalidate(path);
+    }
+  }
+
+  /**
+   * Invalidates the listing cache of its parent of a given path.
+   * If root is specified, the listing cache of root itself will be invalidated.
+   * @param path the full ufs path
+   */
+  public void invalidateListingCacheOfParent(String path) {
+    if (mListStatusCache != null) {
+      mListStatusCache.invalidate(getPathParent(path));
+    }
+  }
+
+  /**
+   * Get the cached listing result from the listing cache.
+   * @param path the full ufs path to list
+   * @param isRecursive if the list is recursive
+   * @return an Optional of a listStatusResult object. If the object exists but the ufs status
+   * is empty, it means that the directory does not exist.
+   */
+  public Optional<ListStatusResult> listCached(String path, boolean isRecursive) {
+    if (mListStatusCache == null) {
+      return Optional.empty();
+    }
+    // We don't cache recursive listing result as usually the number of files are too
+    // large to cache.
+    if (isRecursive) {
+      return Optional.empty();
+    }
+    ListStatusResult result = mListStatusCache.getIfPresent(path);
+    return Optional.ofNullable(result);
+  }
+
+  /**
+   * Lists a directory from UFS and cache it into the listing cache if it exists.
+   * @param path the ufs path
+   * @param isRecursive if the listing is recursive
+   * @return an empty option if the directory does not exist or
+   * the path does not denote a directory,
+   * otherwise an option contains an ufs status array.
+   * @throws IOException if the UFS call failed
+   */
+  public Optional<UfsStatus[]> listFromUfsThenCache(String path, boolean isRecursive)
+      throws IOException {
+    if (mListStatusCache == null) {
+      return listFromUfs(path, isRecursive);
+    }
+    try {
+      var cached = mListStatusCache.get(path, (k) -> {
```

Comment:
`var` is a new syntax introduced by JDK11?
Can you avoid this? Or you prefer this way?

Reply:
yeah i switched it back